### DISCUSSION TITLE
Meaning of TaggingInterface

### DISCUSSION THREAD

**Discussion by matthiasneuner** [idaholab/moose#21855] [Q&A General] [answered]
Hi, I would like to understand the purpose and the function of the `TaggingInterface`.

What is the purpose and the meaning? What is a tag? Why is there a need for a tag? Unfortunately, the doxygen documentation is not very helpful in this regard ..
Also, I cannot really understand how it works.

To me it seems that `_re_blocks` and `_ke_blocks` are populated by kernels, but I cannot see at which stage and by which module those attributes are accessed. That means, I cannot see when and where does attributes are accessed during the final assembly of the equation system.

When should the `TaggingInterface` be used, and why are some modules (e.g., `NodalConstraint`) not using it?

Thank you in advance!

**Comment by GiudGiud** [comment, marked as answer]
Hello

My understanding is that it's meant for two things:
- easy storage and retrieval of vectors (fields) in the system class. I used vector tags for saving the iterates in PicardSolve and SecantSolve
- relatedly, a way to store the contribution of a single kernel to a residual to examine it for debugging

Some modules are not using it because they predate the interface most likely, and there was not necessarily funding to take on a refactor. 

Guillaume
